### Model training notebook

In [1]:
# Dataset
from pandas import read_csv

# Machine Learning
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score

# Sklearn intel extension
from sklearnex import patch_sklearn
patch_sklearn()

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


In [14]:
# Read the training data
training_df = read_csv('../data/final_datasets/transformed_datasets/transformed_reduced_smote_training_data.csv')

# Read the test data
test_df = read_csv('../data/final_datasets/transformed_datasets/transformed_reduced_test_data.csv')

# Read the submission data
submission_df = read_csv('../data/initial_datasets/Submission_template_UHI2025-v2.csv')

In [15]:
# Separate predictors and responses
X = training_df.drop(columns = ['UHI Index binned']).values
y = training_df['UHI Index binned'].values

In [16]:
# Separate predictors and responses
X_test = test_df.values

In [17]:
# Train the Random Forest model on the training data
model = RandomForestRegressor(n_estimators = 128, 
                              ccp_alpha = 2.5e-7, 
                              max_features = 0.3, 
                              min_samples_leaf = 0.01, 
                              max_depth = 12,
                              random_state = 10)
model.fit(X, y)

RandomForestRegressor(ccp_alpha=2.5e-07, max_depth=12, max_features=0.3,
                      min_samples_leaf=0.01, n_estimators=128, random_state=10)

In [ ]:
#df_names = read_csv('../data/final_datasets/raw_datasets/raw_reduced_training_data.csv')
#corr = df_names.corr(method = 'spearman')

In [ ]:
#for column, correlation, importance in zip(df_names.columns[:-1], corr['UHI Index'].values[:-1], model.feature_importances_):
#    print(column, correlation, importance)

gndvi_median_res10 -0.23242072169639502 0.0012227635738696213
datt1_median_res10 -0.10112222266059069 2.485425341878366e-05
fs_median_res10 0.23669203740624217 0.0012456585287809673
siwsi_median_res10 -0.23756384399757405 0.0005706218251612701
ndmi_median_res10 -0.19692658420060125 0.00011341896021034898
si_median_res10 0.14768744061173628 0.00011926092638620328
w_median_res10 -0.13322188268910745 0.0002423806361040697
evi_median_res10 -0.20325645482723342 0.0008700627865158615
B2_median_res10 0.17331527364580235 0.00021308709104181904
B3_median_res10 0.1707866849907093 0.00010887453369194012
B4_median_res10 0.1721538369503773 0.0001464917834489188
B5_median_res10 0.16861792634929842 0.00029904991906469015
B7_median_res10 -0.10985123255260303 0.002016448714526628
B8_median_res10 -0.1322424711376458 0.0011923735023967683
B8A_median_res10 -0.12267888198808166 0.0016258476086146421
B11_median_res10 0.10195918314314867 0.000539436646629188
B12_median_res10 0.18389258468059266 0.00081457638

In [18]:
# Make predictions on the training data
insample_predictions = model.predict(X)

In [19]:
# calculate R-squared score for in-sample predictions
y_train = y.tolist()
r_squared = r2_score(y_train, insample_predictions)
print(r_squared)

0.8515735852167875


- Submission 1: (full dataset) training $R^2 = 0.9863$
- Submission 2: (reduced dataset) training $R^2 = 0.9822$
- Submission 3: (reduced SMOTE dataset) training $R^2 = 0.9921$
- Submission 4: (reduced SMOTE dataset - n_stimators = 50) training $R^2 = 0.9915$
- Submission 5: (reduced dataset - n_stimators = 50) training $R^2 = 0.9811$
- Submission 6: (reduced dataset - n_estimators = 128, ccp_alpha = 5.0e-7, max_features = 0.3, min_samples_leaf = 0.01, max_depth = 8) training $R^2 = 0.3278$
- Submission 7: (reduced SMOTE dataset - n_estimators = 128, ccp_alpha = 5.0e-7, max_features = 0.3, min_samples_leaf = 0.01, max_depth = 8) training $R^2 = 0.6184$
- Submission 8: (complete reduced dataset - n_estimators = 128, ccp_alpha = 2.5e-7, max_features = 0.3, min_samples_leaf = 0.01, max_depth = 11) training $R^2 = 0.6834$
- Submission 9: (complete reduced SMOTE dataset - n_estimators = 128, ccp_alpha = 2.5e-7, max_features = 0.3, min_samples_leaf = 0.01, max_depth = 12) training $R^2 = 0.8516$

In [20]:
# Make predictions on the test data
predictions = model.predict(X_test)

In [21]:
# Replace predictions of test data on submission dataset
submission_df['UHI Index'] = predictions

In [22]:
# Save the submission dataset
submission_df.to_csv('../data/submissions/submission-v9.csv', index = False)